core/src/main/java/org/elasticsearch/cluster/routing/allocation/IndexMetaDataUpdater.java

```diff
@@ -39,23 +44,23 @@
  * Observer that tracks changes made to RoutingNodes in order to update the primary terms and in-sync allocation ids in
  * {@link IndexMetaData} once the allocation round has completed.
  *
- * Primary terms are updated on primary initialization or primary promotion.
+ * Primary terms are updated on primary initialization or when an active primary fails.
  *
  * Allocation ids are added for shards that become active and removed for shards that stop being active.
  */
 public class IndexMetaDataUpdater extends RoutingChangesObserver.AbstractRoutingChangesObserver {
     private final Map<ShardId, Updates> shardChanges = new HashMap<>();
 
     @Override
-    public void shardInitialized(ShardRouting unassignedShard) {
-        if (unassignedShard.primary()) {
-            increasePrimaryTerm(unassignedShard);
-        }
-    }
+    public void shardInitialized(ShardRouting unassignedShard, ShardRouting initializingShard) {
```

Comment:
I've put both there as the interface is then consistent with the other methods in the interface. `shardStarted` e.g. has as parameter the initialized shard as well as the started shard. Same for other methods.